pkg/storage/prunable/epochstore/epoch_kv.go

```diff
@@ -0,0 +1,79 @@
+package epochstore
+
+import (
+	"github.com/iotaledger/hive.go/ierrors"
+	"github.com/iotaledger/hive.go/kvstore"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/iota-core/pkg/model"
+	"github.com/iotaledger/iota-core/pkg/storage/database"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+type EpochKVStore struct {
+	realm        kvstore.Realm
+	kv           kvstore.KVStore
+	pruningDelay iotago.EpochIndex
+
+	lastPrunedEpoch *model.PruningIndex
+}
+
+func NewEpochKVStore(storeRealm, pruningRealm kvstore.Realm, kv kvstore.KVStore, pruningDelay iotago.EpochIndex) *EpochKVStore {
+
+	return &EpochKVStore{
+		realm:           storeRealm,
+		kv:              lo.PanicOnErr(kv.WithExtendedRealm(storeRealm)),
+		pruningDelay:    pruningDelay,
+		lastPrunedEpoch: model.NewPruningIndex(lo.PanicOnErr(kv.WithExtendedRealm(pruningRealm)), storeRealm),
+	}
+}
+
+func (e *EpochKVStore) isTooOld(epoch iotago.EpochIndex) bool {
+	prunedEpoch, hasPruned := e.lastPrunedEpoch.Index()
+
+	return hasPruned && epoch <= prunedEpoch
+}
+
+func (e *EpochKVStore) RestoreLastPrunedEpoch() error {
+	return e.lastPrunedEpoch.RestoreFromDisk()
+}
+
+func (e *EpochKVStore) LastPrunedEpoch() (iotago.EpochIndex, bool) {
+	return e.lastPrunedEpoch.Index()
+}
+
+func (e *EpochKVStore) GetEpoch(epoch iotago.EpochIndex) (kvstore.KVStore, error) {
+	if e.isTooOld(epoch) {
+		return nil, ierrors.Wrapf(database.ErrEpochPruned, "epoch %d is too old", epoch)
+	}
+
+	return lo.PanicOnErr(e.kv.WithExtendedRealm(epoch.MustBytes())), nil
+}
+
+func (e *EpochKVStore) Prune(epoch iotago.EpochIndex, defaultPruningDelay iotago.EpochIndex) error {
+	// The epoch we're trying to prune already takes into account the defaultPruningDelay.
+	// Therefore, we don't need to do anything if it is greater equal e.pruningDelay and take the difference otherwise.
+	var pruningDelay iotago.EpochIndex
```

Comment:
I don't really understand this logic - could you explain a little more?

Reply:
This is indeed a bit tricky and took us a while to figure out properly. 

Let's assume you could prune epoch 10 (which means you have at least finalized a slot within epoch 12). Now, instead of directly pruning epoch 10, you want to keep always 5 epochs = `defaultPruningDelay`. That means you will prune until epoch 5. 
Because all of the epoch stores have their own pruning delay you need to take the max out of both. For example, for the `decidedUpgradeSignals` it's 7. So you want to prune _that_ storage until epoch `10-7=3`. The problem is, that the epoch you'll prune by default (meaning you drop the buckets and delete the spent ledger) already "contains" the `defaultPruningDelay` of `10-5=5`. That's why you can't simply do `5-7=-2->0 and don't prune` but instead need to take the difference.